### ISSUE TITLE
Customize LED segment display

### ISSUE THREAD

**Issue by MattKunze** [UltimateHackingKeyboard/agent#660]
Currently the LED segments show the currently selected keymap ('QWM', etc). For me at least this is something that will never change, so the display doesn't really provide much information. It would be awesome if these could be overridden with a console utility (`uhk -leds 'FOO'`) to display arbitrary text.

Why? This is the ultimate *hacking* keyboard, so mainly it just seems like stuff like that should be hackable ;) I could see writing a script to update it with the current CPU load, polling the outside temperature, hooking it up to the current terminal to display additional status, who knows.

Not sure if this belongs here or the firmware repo, thought I'd get the discussion started somewhere though.

**Comment by mondalaci** [comment]
Good point, and this is something that we planned. Now we also have an issue for it. Will do eventually. :)

**Comment by luteijn** [comment]
To expose the LED (and future expansions, like e.g. an i2c lcd display or other gadget) it might be useful to have the uhk also present itself as a usb-serial converter so scripts have something straight-forward to talk to. Alternatively/additionally agent could act as a proxy, but I guess you already have some ideas for this.

**Comment by mondalaci** [comment]
@luteijn Yep, the plan is to expose Agent as a proxy. Regarding communication, the already existing generic USB HID interface will be used. It's binary, but creating an additional USB serial interface would create a lot of bulk.

**Comment by mondalaci** [comment]
@richrd Yep, the LED display segments will be individually controllable. This is much needed for showing characters that resemble the predator alphabet, and who are we to stop you from doing that. :)

So glad you're having fun with your UHK!

**Comment by lrascao** [comment]
👍 i would also like to see this, btw great job!

**Comment by Spark3000** [comment]
Awesome.

**Comment by norrs** [comment]
This would be awesome, could use it for showing current xkbkeymap language, ie if Im currently using english keyboard or native norwegian.

**Comment by FatBoyXPC** [comment]
I think it would be neat if we could include some scripts to be stored on the keyboard, triggerable by the keyboard as well. Similar to how you can change layouts with the keyboard.

**Comment by mondalaci** [comment]
@FatBoyXPC This is not planned. Generally speaking, Agent is only meant to configure the UHK. You can achieve what you're after by setting up an unused shortcut (such as Ctrl+Alt+Shift+F13) for a key and bind it to your window manager to execute a script.

**Comment by FatBoyXPC** [comment]
Yeah, I figured that would be the immediate implementation but I was hoping having scripts there were movable with the keyboard would be on road map, too. It's not a big deal, though, as really my main use case for this is only with one computer anyway.

**Comment by jsarbour** [comment]
+1 to this FR. I'd love to use mine as a clock (and run the hours in Hexadecimal I guess on one digit)

**Comment by BrianGaucher** [comment]
Using it as a clock would be great. What if there was a LED display module?

**Comment by BrianGaucher** [comment]
Using it as a clock would be great. What if there was a LED display module?

**Comment by kareltucek** [comment]
Cmon, this is a keyboard, not a second display. If you need second display, get a second display. It will be cheaper.

**Comment by bisgaardmichael** [comment]
I also want this feature, since I want to have different macros for my vim for c++ and LaTeX and other languages I might adopt. And it would be nice too see which one was the active one.

**Comment by mhantsch** [comment]
The idea of an UHK clock inspired me: there are three positions in the display. If you consider only the segments pointing outwards from the center, you have 8 of those in 360 degrees. Does not work well for advanced the "hand" every hour, or every minute, but try this:

![image](https://user-images.githubusercontent.com/591785/84050274-cb17ea80-a9ad-11ea-98eb-f2dae3786b47.png)

**Comment by kareltucek** [comment]
...I guess you could also make the brightness change independently per segment and make it "continuous"...

**Comment by mondalaci** [comment]
All of this will be possible as the individual segments and per-key backlights will be exposed via the UHK USB protocol. Writing must be possible by the host, but I'm wondering whether reading LED states would make sense.

**Comment by csmote** [comment]
Realize this isn't a priority, but just something to get me started here.

As a (hopefully) soon, to be owner of an UHK, I'd love to see the LED show the current "modifier" level (for those that aren't already shown in the images in the same area), i.e., MOD, FN, CTL, ALT, SUP, CMD, WIN,  MOU, etc. and also love to be able to have those flash and control the rate of flash per code displayed. Particularly useful for getting a visual reminder to switch out of tenkey or arrow mode (especially arrow mode since that's going to be used often - I'd even have that flash all segs rapidly to really have a "warning" that not in standard key board mode.

Such a great piece of engineering. Keeping an eye on my resources to get an order together.

**Comment by mhantsch** [comment]
@csmote I configured my NumPad "layer" not as a layer, but as a keymap. When I switch to it, the LED display shows that I am on the NumPad map. I guess that's a little bit of what you wanted (although only for NumPad, not for all the other modifiers).

**Comment by csmote** [comment]
@mhantsch. Thanks for replying with your work-around - very slick! Understanding that the UHK is an evolving platform that's made tremendous progress so far!

**Comment by stn1986** [comment]
While I don't actually own the keyboard yet (just ordered it) I was also wondering about this. I'd love for this tiny display to tell me the status of some of my projects, servers, emergencies, that sort of thing. I'm kinda bummed this LED is not directly programmable, but it's not a deal breaker. Anyway:

I see on YT you can create custom keymaps and name them, which is then displayed on the LED when activated. I only ever use one keymap, so my idea is to create a bunch of copies of that map, name them differently and switch them externally when some event is triggered.

Let's say I use the three digits as code. 000 = no problems, 100 = project 1 is in trouble, 010 = project 2 is in trouble and 003 = project 3 is in trouble. I create four duplicate keymaps that are exactly the same except for the name ("000", "100", etc). Let's assume I have some sort of code running externally that detects these problems and can run the same code the actual keyboard software uses to switch keymaps. In that case I can set the display to 000 and 100 and so on by automatically switching keymaps (which are actually the same so it doesn't impact my typing), right?

Let me know if I'm talking out of my ass here. I'm a developer so I don't mind the code, but perhaps I'm fundamentally misunderstanding something else here.

PS: I see https://github.com/UltimateHackingKeyboard/agent/blob/4ecc3ef795e8d13b2ef2acead9066899f087b5c3/packages/uhk-usb/src/uhk-operations.ts#L447 is the code to switch keymaps. The codebase looks really clean by the way. Real easy to navigate.

**Comment by kareltucek** [comment]
> Let me know if I'm talking out of my ass here. 

In theory, it works. In practice, it would be extremely impractical.

Since you can code, just add new usb command similar to the referenced `switchKeymap`, and set the LED display to the desired text. It will not change until something else changes it - in practice, until next keymap switch.

Some references:
- `agent/packages/usb/switch-keymap.ts`
- in firmware: `LedDisplay_SetText`,
- `right/src/usb_commands/usb_command_switch_keymap.c`, or maybe more relevant: the set variable command.

**Comment by stn1986** [comment]
I was indeed wondering if I could write arbitrary strings there, but I assumed the device must know the name.

What is impractical about it? It sounds like I directly write any string to the display.

**Comment by stn1986** [comment]
Thanks @kareltucek !

**Comment by kareltucek** [comment]
You are welcome!

**Comment by stn1986** [comment]
So sorry to keep hammering on this, but I see in the FW in led_display.c that I can control the segments individually, which is.. making me giddy like a little girl. I'm not familiar with "the scene" so to speak, but is there a tool or something already written by someone to control the display directly? To me it'd be really cool to do what I want with it. I guess I could be of service to the community here if I'd provide such a tool until the UHK teams releases their own implementation. If someone did not beat me to it that is, which would be awesome as well.

**Comment by kareltucek** [comment]
> but is there a tool or something already written by someone to control the display directly?

I am not aware of it.

> I guess I could be of service to the community here if I'd provide such a tool until the UHK teams releases their own implementation. If someone did not beat me to it that is, which would be awesome as well.

I am sure a PR will be welcome, just consult your API proposal with @mondalaci to make sure you are on the same page.

**Comment by stn1986** [comment]
OK, cool. I'll mess around with it first when I get mine and I'll look at your fork. Thanks again.

**Comment by maxbethke** [comment]
Just wanted to add my +1 to this and drop some thoughts.
I just got the UHK v2 and this is the only thing that puts me off about it.

As stated above, the keymap would rarely change for the most users, given that there are 10+ layers on each keymap, so I see no need to sacrifice the display for that info.

I fantasized about being able to toggle the state of the display using the modifier keys, so for example, in normal mode the display would show the current time, in mod the CPU Temperature, in fn the current value of the bitcoin and on mouse mode time left in my current pomodoro cycle or anything else.

**Comment by mondalaci** [comment]
@maxbethke Your use cases are interesting, and I'd like to support them, but we'll need to expose a sophisticated API to enable them.

1. Currently, the `setLedTxt` smart macro command enables the override of the LED display text.
2. https://github.com/UltimateHackingKeyboard/firmware/issues/531 will enable the execution of smart macros, such as `setLedTxt` via USB.
3. We'll expose a nice TypeScript API on top of the USB commands to ease implementing custom behavior, as the one suggested. Currently, `${agentDirectory}/packages/usb` contains some random scripts that will serve as the basis of such an API.
4. We'll expose a dedicated USB command to set per-LED brightness (including the LED display and per-key LEDs) efficiently.

**Comment by arcxio** [comment]
> This would be awesome, could use it for showing current xkbkeymap language, ie if Im currently using english keyboard or native norwegian.

for anyone using sway, I made a script that does this
```shell
#!/bin/sh
swaymsg -m -t subscribe '["input"]' |
    jq --unbuffered -r '.|select(.change == "xkb_layout")|.input.xkb_active_layout_name' | (

    prev_layout=""
    while read new_layout; do
        if [ "$new_layout" != "$prev_layout" ]; then
            prev_layout=$new_layout
            hidraw=$(grep 'UHK 60' /sys/class/hidraw/hidraw*/device/uevent | LC_ALL=C sort -rh | head -n 1 | grep -o 'hidraw[0-9][0-9]*')
            printf "\x14setLedTxt 0 %s" "$(printf %s "$new_layout" | cut -c-3)" > "/dev/$hidraw"
        fi
    done
)
```
